packages/blade/src/components/Input/DropdownInputTriggers/AutoComplete.tsx

```diff
@@ -0,0 +1,198 @@
+      setIsOpen(true);
+    }
+
+    if (!props.filteredValues) {
+      // eslint-disable-next-line no-lonely-if
+      if (value && options && options.length > 0) {
+        const filteredOptions = getOptionValues().filter((optionValue) =>
+          optionValue.toLowerCase().startsWith(value.toLowerCase()),
+        );
+        setGlobalFilteredValues(filteredOptions);
+      } else {
+        setGlobalFilteredValues(getOptionValues());
+      }
+    }
+  };
+
+  const onTriggerKeydown: BaseDropdownInputTriggerProps['onTriggerKeydown'] = (e) => {
+    if (e.key === 'Enter') {
+      setActiveTagIndex(-1);
+      setGlobalFilteredValues(getOptionValues());
+    } else if (
+      e.key === 'Backspace' &&
+      !inputValue &&
+      activeTagIndex < 0 &&
+      selectedIndices.length > 0
+    ) {
+      if (isControlled) {
+        setControlledValueIndices(selectedIndices.slice(0, -1));
+      } else {
+        setSelectedIndices(selectedIndices.slice(0, -1));
+      }
+    }
+    onBaseDropdownInputKeydown?.(e);
+  };
+
+  const onSelectionChange: BaseDropdownInputTriggerProps['onChange'] = ({ values }) => {
+    if (selectionType === 'multiple') {
+      setInputValue('');
+      props.onInputValueChange?.({ name: props.name, value: '' });
+      setActiveTagIndex(-1);
+      setGlobalFilteredValues(getOptionValues());
+    } else {
+      const displayText = options.find((option) => option.value === values[0])?.title;
+      props.onInputValueChange?.({
+        name: props.name,
+        value: displayText,
+      });
+      setInputValue(displayText ?? '');
+    }
+    props.onChange?.({ name: props.name, values });
+  };
+
+  return {
+    onSelectionChange,
+    onTriggerKeydown,
+    onInputValueChange,
+  };
+};
+
+const _AutoComplete = (
+  props: AutoCompleteProps,
+  ref: React.ForwardedRef<BladeElementRef>,
+): React.ReactElement => {
+  const [uncontrolledInputValue, setInputValue] = React.useState('');
+  const inputValue = props.inputValue ?? uncontrolledInputValue;
+
+  const {
+    isOpen,
+    options,
+    setFilteredValues: setGlobalFilteredValues,
+    hasAutoCompleteInBottomSheetHeader,
+    setHasAutoCompleteInBottomSheetHeader,
+    onTriggerClick,
+    dropdownTriggerer,
+  } = useDropdown();
+
+  const getOptionValues = React.useCallback(() => {
+    return options.map((option) => option.value);
+  }, [options]);
+
+  const { onSelectionChange, onTriggerKeydown, onInputValueChange } = useAutoCompleteHandlers({
+    props,
+    inputValue,
+    setInputValue,
+    getOptionValues,
+  });
+
+  React.useEffect(() => {
+    if (dropdownTriggerer !== 'AutoComplete') {
+      // When AutoComplete is mounted but not as trigger,
+      // we assume its in header of BottomSheet
+      setHasAutoCompleteInBottomSheetHeader(true);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  React.useEffect(() => {
+    if (isOpen && !inputValue) {
+      setGlobalFilteredValues(getOptionValues());
```

Comment:
We're doing this very often and in multiple places. I'm still not sure what we're doing with this and if its safe to do it so many times 😅

Reply:
Yup realised that as well. It is safe but will check if I can refactor this into separate hook maybe

Reply:
moved some of it inside useAutoComplete hook. Also added comment for each useEffect block. Don't think so we can remove any of them as they are all doing something expected.